blocks_vertical/procedures.js

```diff
@@ -312,91 +345,208 @@ Blockly.ScratchBlocks.ProcedureUtils.attachArgumentReporter_ = function(
  * TODO (#1213) consider renaming.
  * @param {string} type One of 'b' (boolean), 's' (string) or 'n' (number).
- * @param {string} name The name to use when adding the input to the block.
  * @param {number} index The index of this input into the argument id and name
  *     arrays.
  * @param {!Object.<string, Blockly.Block>} connectionMap An object mapping
  *     parameter IDs to the blocks that were connected to those IDs at the
  *     beginning of the mutation.
+ * @param {Blockly.BlockSvg} oldBlock The block that was previously connected to
+ *     this input, or null.
+ * @param {!Blockly.Input} input The newly created input to populate.
  * @private
  * @this Blockly.Block
  */
-Blockly.ScratchBlocks.ProcedureUtils.createInputCallerInternal_ = function(type,
-    name, index, connectionMap) {
-  var id = this.argumentIds_[index];
-  var oldBlock = null;
-  if (connectionMap && (id in connectionMap)) {
-    oldBlock = connectionMap[id];
+Blockly.ScratchBlocks.ProcedureUtils.populateInputCallerInternal_ = function(type,
+    index, connectionMap, oldBlock, input) {
+  var oldTypeMatches =
+    Blockly.ScratchBlocks.ProcedureUtils.checkOldTypeMatches_(oldBlock, type);
+  var displayName = this.displayNames_[index];
+  if (connectionMap && oldBlock && oldTypeMatches) {
+    // Reattach the old block, and update the text if needed.
+    // The old block is the same type, and on the same input, but the input name
+    // may have changed.
+    oldBlock.setFieldValue(displayName, 'VALUE');
+    connectionMap[input.name] = null;
+    oldBlock.outputConnection.connect(input.connection);
+  } else {
+    this.attachArgumentReporter_(input, type, displayName);
   }
-  var input = this.appendValueInput(name);
-  if (connectionMap && oldBlock) {
-    this.reattachBlock_(input, type, oldBlock, id, connectionMap);
+};
+
+/**
+ * Populate the given input with the correct child block or values.
+ * This function is used by the procedures_callnoreturn_internal block.
+ * TODO (#1213) consider renaming.
+ * @param {string} type One of 'b' (boolean), 's' (string) or 'n' (number).
+ * @param {number} index The index of this input into the argument id and name
+ *     arrays.
+ * @param {!Object.<string, Blockly.Block>} connectionMap An object mapping
+ *     parameter IDs to the blocks that were connected to those IDs at the
+ *     beginning of the mutation.
+ * @param {Blockly.BlockSvg} oldBlock The block that was previously connected to
+ *     this input, or null.
+ * @param {!Blockly.Input} input The newly created input to populate.
+ * @private
+ * @this Blockly.Block
+ */
+Blockly.ScratchBlocks.ProcedureUtils.populateInputDeclarationRoot_ = function(type,
+    index, connectionMap, oldBlock, input) {
+  var oldTypeMatches =
+    Blockly.ScratchBlocks.ProcedureUtils.checkOldTypeMatches_(oldBlock, type);
+
+  var displayName = this.displayNames_[index];
+  if (connectionMap && oldBlock && oldTypeMatches) {
```

Comment:
Is this code essentially the same as what's in populateInputCallerInternal_ except that setFieldValue gets called with different type?